cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/handler/RuntimeHandler.java

```diff
@@ -0,0 +1,135 @@
+import javax.ws.rs.Path;
+
+/**
+ * {@link co.cask.http.HttpHandler} for exposing metadata of a runtime.
+ */
+@Path("/v1/runtime")
+public class RuntimeHandler extends AbstractHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_CONSUME_REQUEST_TYPE = new TypeToken<Map<String,
+    MonitorConsumeRequest>>() { }.getType();
+  private static final int CHUNK_SIZE = 8192;
+
+  private final CConfiguration cConf;
+  private final MessageFetcher messageFetcher;
+  private final Runnable shutdownRunnable;
+
+  public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runnable shutdownRunnable) {
+    this.cConf = cConf;
+    this.messageFetcher = messageFetcher;
+    this.shutdownRunnable = shutdownRunnable;
+  }
+
+  /**
+   * Gets list of topics along with offsets and limit as request and returns list of messages
+   */
+  @POST
+  @Path("/metadata")
+  public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
+    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+
+    Map<String, MonitorConsumeRequest> consumeRequests = GSON.fromJson(requestBody, MAP_STRING_CONSUME_REQUEST_TYPE);
+
+    ChunkResponder chunkResponder = responder.sendChunkStart(
+      HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE,
+                                                          "application/json; charset=utf-8"));
+
+    ByteBuf buffer = Unpooled.buffer();
+    JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
+                                                                  StandardCharsets.UTF_8));
+    jsonWriter.beginObject();
+
+    for (Map.Entry<String, MonitorConsumeRequest> entry : consumeRequests.entrySet()) {
+      jsonWriter.name(entry.getKey());
+      jsonWriter.beginArray();
+
+      writeMessages(jsonWriter, buffer, chunkResponder, cConf.get(entry.getKey()), entry.getValue().getLimit(),
+                    entry.getValue().getMessageId());
+
+      jsonWriter.endArray();
+    }
+
+    jsonWriter.endObject();
+    jsonWriter.close();
+
+    if (buffer.isReadable()) {
+      chunkResponder.sendChunk(buffer.copy());
+    }
+
+    Closeables.closeQuietly(chunkResponder);
+  }
+
+  @POST
+  @Path("/shutdown")
+  public void shutdown(FullHttpRequest request, HttpResponder responder) throws Exception {
+    responder.sendString(HttpResponseStatus.OK, "Triggering shutdown down Runtime Http Server.");
+    shutdownRunnable.run();
+  }
+
+  private void writeMessages(JsonWriter jsonWriter, ByteBuf buffer, ChunkResponder chunkResponder, String topic,
+                             int limit, @Nullable String fromMessage) throws TopicNotFoundException, IOException {
+    try (CloseableIterator<Message> iter = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(), topic, limit,
```

Comment:
So if there is exception, then the resulting json object would be broken?

Reply:
In that case the response will be a string: https://github.com/caskdata/cdap/blob/develop/cdap-common/src/main/java/co/cask/cdap/common/HttpExceptionHandler.java#L72

The Client should handle this case. I will modify the client accordingly.

Reply:
It won't work.. You already sent the partial json, how can it be a string then?

Reply:
How is it on the client side?? Imagine that you have outputted this so far:

```
{
  "program.state.topic" : [
    { "messageId" : "xyz", "message" : "RUNNING" },
```

Then on the next message in the loop, you got an exception. If you just let the exception propagate, then the client will get the broken json with a 200 response code. Do you expect the client to catch that?

I think it's better to handle it on the server side to always have a valid json response. If there is exception, you can catch it and close the json object appropriately..